Extension marketplace

maifeeulasad opened the pull request:
The UI looks something like this now

<img width="812" height="1250" alt="freelens getting extension marketplace, by maifeeulasad" src="https://github.com/user-attachments/assets/48b276a3-ec70-4995-a394-b1b96a8094fd" />

Installed extensions UI has also been updated! I am adding two sample capture with mock data:

<img width="812" height="1250" alt="freelens updated installed extension list ui, by maifeeulasad" src="https://github.com/user-attachments/assets/1fe324d0-1d43-48fd-998a-58a41a1bc429" />
<img width="812" height="1250" alt="freelens updated installed extension list ui, by maifeeulasad" src="https://github.com/user-attachments/assets/d4efb800-10b8-4c1a-8b56-45e67ae9936a" />

closes #730

maifeeulasad committed:
[ui]: new extension card; extension grid; search bar for extension;

maifeeulasad committed:
[feat]: marketplace extension based on new ui;

maifeeulasad committed:
[feat]: installed extension based on new ui;

maifeeulasad committed:
[ui]: new extension layout w installed and marketplace ui;

maifeeulasad committed:
[lint]: ran `pnpm biome:fix && pnpm prettier:fix && pnpm trunk:fix && pnpm trunk:fix:all && pnpm lint:fix`

mariomamo reviewed (changes requested):
It is ok for me, just avoid to hardcode css and create module.scss file

maifeeulasad commented:
@mariomamo I will modularize the whole style into a dedicated stylesheet next week. Does this work?

maifeeulasad committed:
[refactor]: extracted style in detecated search bar module scss;

maifeeulasad committed:
[refactor]: extracted style in detecated extension grid module scss;

maifeeulasad committed:
[refactor]: extracted style in detecated extension card module scss;

maifeeulasad committed:
[refactor]: extracted style in detecated extension module scss;

maifeeulasad committed:
[refactor]: extracted style in detecated marketplace extension module scss;

maifeeulasad committed:
[lint]: ran `pnpm biome:fix && pnpm prettier:fix && pnpm trunk:fix && pnpm trunk:fix:all && pnpm lint:fix`

maifeeulasad committed:
Merge branch 'freelensapp:main' into extension-marketplace-maifee

maifeeulasad committed:
[fix]: generated regster injectables w `pnpm build:di`;

maifeeulasad committed:
[test]: updated snap w `pnpm test:unit:core:updatesnapshot && pnpm test:unit:updatesnapshot`;

maifeeulasad committed:
[test,fix]: uninstall test with menu trigger handling; diffrent state assertions;

maifeeulasad committed:
Merge branch 'main' into extension-marketplace-maifee

mariomamo commented:
Nice work! The last part is to edit `marketplace-extensions.injectable.ts` so that it calls the Artifact Hub APIs to get the list of extensions.  
Do you confirm that this is the structure we need? If so, I can open an issue on Artifact Hub to create a kind for our use case.  

```json
  {
    "id": "freelens-ai",
    "name": "@freelensapp/ai-extension",
    "description": "Freelens AI",
    "version": "0.4.1",
    "author": "freelens",
    "downloads": 0,
  },
```

maifeeulasad commented:
For now this seems okay, definetly icon will be a great addition.

On top of that, I think we need kubernates version and other metadata, as for some version of externsion won't be compatible with certain kubernates. So I think we will be needing those as well, going forward. I haven't thought about those completly yet. So if you have any input feel free to share. I am thinking out loud, so that we can make the foundation stronger.

robertobandini commented:
We've talked about this briefly within the Freelens core team and release team.
At the moment, it seems like the best solution is to have a dedicated repo where we can start with a single JSON file containing metadata for both "official" and third-party extensions.
Anyone can submit PR to be included.
We can discuss how to structure the repo: whether to start with just a single, hand-written file or have a structure that facilitates changes for each extension with automatic workflows that update the JSON file.
Eventually, as a next step, we could also consider using GitHub Pages on the dedicated repo to create a web interface to search for extensions in the JSON file from the UI, which will be the same thing that can be done from Freelens as a desktop app in its extensions section.
If everyone is okay with that, I can already create the repo. Is "freelens-marketplace" okay?
This way, @mariomamo  and @maifeeulasad  can continue and direct the development of this part, with the others reviewing.
As always, anyone's opinions/ideas are always welcome, thank you.

@dex4er @MatiasRoje @omarluq

maifeeulasad committed:
[feat]: list extensions from marketplace;
 - hosted in github repo

maifeeulasad committed:
[refactor]: fetch marketplace extension list over static mock data;

maifeeulasad committed:
[fix]: generated regster injectables w `pnpm build:di`;

maifeeulasad committed:
[feat]: install extension w version from marketplace list;

maifeeulasad committed:
[lint]: ran lint related script;
 - ran `pnpm biome:fix && pnpm prettier:fix && pnpm trunk:fix && pnpm trunk:fix:all && pnpm lint:fix`

maifeeulasad committed:
[refactor]: changed `packages` into `extensions` for marketplace;

maifeeulasad committed:
[fix]: mark community managed extensions;

robertobandini commented:
Guys, I think it would be nice to include this on the 1.9, what do you think?

maifeeulasad commented:
Honestly, I would love to see this in `v1.9`!! Eagerly waiting to see it in production!

robertobandini commented:
@maifeeulasad Can you check the failed tests?

maifeeulasad committed:
Merge branch 'main' into extension-marketplace-maifee

mariomamo commented on packages/core/src/renderer/components/extensions/marketplace-extensions/request-marketplace-extensions.injectable.ts:
```diff
@@ -0,0 +1,97 @@
+// todo: change to offficial url, once the MR is merged
+// ref: https://github.com/freelensapp/freelens-marketplace/pull/1
+const extensionListUrl =
+  "https://raw.githubusercontent.com/maifeeulasad/freelens-marketplace/refs/heads/marketplace-v0/extensions.json";
```
This url should point to freelens repo

maifeeulasad replied:
Yes, as it was mentioned in todo, it was because the other MR was not merged yet.

But I have removed the todo and updated the url.

mariomamo commented on packages/core/src/renderer/components/extensions/marketplace-extensions.tsx:
```diff
@@ -0,0 +1,79 @@
+      (ext) =>
+        ext.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
+        ext.description.toLowerCase().includes(searchQuery.toLowerCase()),
+    );
```
As suggested by Copilot, consider using useMemo to prevent performance performance issues if the extension list grows larger

mariomamo reviewed (changes requested):
The extension json url should be changed to freelens repo

maifeeulasad committed:
refactor: optimize filtered extension for installed extension

maifeeulasad committed:
lint: ran `pnpm biome:fix && pnpm prettier:fix && pnpm trunk:fix && pnpm trunk:fix:all && pnpm lint:fix`

maifeeulasad committed:
fix: change to official url for extension marketplace

maifeeulasad committed:
Merge branch 'main' into extension-marketplace-maifee
